### PR TITLE
Pulsar add-on documentation

### PR DESCRIPTION
**Pull request by Keksoj**
As described in #660, we need to provide good documentation to use the pulsar addon (currently in beta version)

Todos:

- [x] brief presentation
- [x] add-on version
- [x] add-on policies
  - [x] storage
  - [x] retention
  - [x] offload
- [x] how to create an add-on
- [x] how to connect to it
  - [x] biscuit examples (rust, java)
  - [x] pulsarctl usage
  - [x] biscuit attenuation (not a priority) with biscuit-cli
- [x] plan & pricing
- [x] add-on section in the environment reference page

### CODE REVIEW

**Comment by KannarFr** [comment]
Nice job @Keksoj , I will add parts to this PR too.

**Comment by Keksoj** [comment]
@KannarFr yeah, the `pulsarctl` parts would be usefull, you know those better than I do.

**Comment by KannarFr** [comment]
@Keksoj @urcadox @gaelreyrol can you review my commits?

@Keksoj  can you provide a rust & java quick usage in a single file? And what do you mean by "how to create an addon"', you want to do a gif on the console, cli commands? I think it's a good idea, feel free.

**Comment by KannarFr** [comment]
@gaelreyrol can you put contextual alert blocks when you think it's mandatory as a new reader and potentially not aware of how pulsar works?

**Comment by gaelreyrol** [comment]
> @Galimede can we embed the pulsar pricing component in this doc page?

The component first need to be released according to the [CHANGELOG](https://github.com/CleverCloud/clever-components/blob/master/CHANGELOG.md) and then a dedicated [shortcode](https://github.com/CleverCloud/hugo-cc-doc-theme/blob/master/layouts/shortcodes/pricingAddon.html) in the theme.

File: deploy/addon/pulsar.md [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+
+More on this in the [official documentation](https://pulsar.apache.org/docs/en/concepts-messaging/#subscriptions)
+
+A topic is defined this way:
+
+`{persistent|non-persistent}://tenant/namespace/topic`
+
+Tenants and namespaces allow for grouping and subgrouping of topics.
+
+What a pulsar add-on basically is, is a tenant/namespace pair that allows you to create and use as many topics as you wantt following this pattern:
```
Comment (gaelreyrol) [inline deploy/addon/pulsar.md:41]: I find difficult to read that one:
`What a pulsar add-on basically is, is`

Comment (gaelreyrol) [reply deploy/addon/pulsar.md:41]:
```suggestion
What a pulsar add-on basically is, is a tenant/namespace pair that allows you to create and use as many topics as you want following this pattern:
```

File: deploy/addon/pulsar.md [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+
+`{persistent|non-persistent}://user_id/pulsar_addon_id/name_of_your_topic`
+
+## Authorization
+
+Pulsar addon uses (Biscuit for Pulsar)[https://github.com/CleverCloud/biscuit-pulsar/commits/master] implementation which is directly pluggable to the Pulsar authentication and authorization system. Each addon exposes its own biscuit token.
```
Comment (gaelreyrol) [inline deploy/addon/pulsar.md:47]:
```suggestion
Pulsar add-on uses (Biscuit for Pulsar)[https://github.com/CleverCloud/biscuit-pulsar] implementation which is directly pluggable to the Pulsar authentication and authorization system. Each add-on exposes its own biscuit token.
```

File: deploy/addon/pulsar.md [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+shortdesc: A distributed messaging and streaming platform on the publish/subsribe model
+tags:
+- addons
+keywords:
+- pulsar
+- messaging
+- cold storage
+- queue
+- hot storage
+- biscuit
```
Comment (gaelreyrol) [inline deploy/addon/pulsar.md:12]:
```suggestion
- biscuit
- rabbitmq
- kafka
```

File: deploy/addon/pulsar.md [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+```
+client               Pulsar             client
+publishes to   =>    topic      <=  subscribes to
+
+"producer"                           "consumer"
+```
+
+They are several modes of subscription. A consumer may subscribe exclusively, or share the subscription with other consumers.
+
+- Exclusive
```
Comment (gaelreyrol) [inline deploy/addon/pulsar.md:28]: What is this list ?

File: deploy/addon/pulsar.md [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+
+`{persistent|non-persistent}://user_id/pulsar_addon_id/name_of_your_topic`
+
+## Authorization
+
+Pulsar addon uses (Biscuit for Pulsar)[https://github.com/CleverCloud/biscuit-pulsar/commits/master] implementation which is directly pluggable to the Pulsar authentication and authorization system. Each addon exposes its own biscuit token.
+
+### Usage
+
+We advise you to use (`pulsarctl`)[https://github.com/streamnative/pulsarctl] provided by StreamNative. Here is an example to list topics in your addon (in your namespace):
```
Comment (gaelreyrol) [inline deploy/addon/pulsar.md:51]:
```suggestion
We advise you to use (`pulsarctl`)[https://github.com/streamnative/pulsarctl] provided by StreamNative. Here is an example to list topics in your add-on (in your namespace):
```

File: deploy/addon/pulsar.md [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+
+```bash
+pulsarctl --admin-service-url https://<PULSAR_REST_URL> --auth-params <BISCUIT> --auth-plugin org.apache.pulsar.client.impl.auth.AuthenticationToken topics list <tenant>/<namespace>
```
Comment (gaelreyrol) [inline deploy/addon/pulsar.md:54]: You should use the add-on environment variables.

File: deploy/addon/pulsar.md [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+```
+
+As Biscuit is a token, you can use `AuthenticationToken("<BISCUIT>")` provided by clients librairies to authenticate to our clusters without any tweak.
+
+* C++ client
```
Comment (gaelreyrol) [inline deploy/addon/pulsar.md:59]: I think should simply link this page: https://pulsar.apache.org/docs/en/client-libraries/

File: deploy/addon/pulsar.md [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+* Go client
+* Java client
+* Node.js client
+* Python client
+* Rust client
+* WebSocket client
+
+### Operations
+
+The pulsar addon given biscuit enable you to run several operations on namespace, its policies and the related topics.
```
Comment (gaelreyrol) [inline deploy/addon/pulsar.md:70]:
```suggestion
The pulsar add-on given biscuit enable you to run several operations on namespace, its policies and the related topics.
```

File: deploy/addon/pulsar.md [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+Topics policies operations authorized:
+
+```
+PARTITION_READ
+PARTITION_WRITE
+```
+
+### Attenuation
+
+The pulsar addon given biscuit can be attenuated, here is an attenuation example using (biscuit-cli)[https://github.com/biscuit-auth/biscuit-cli] from the given biscuit to produce/consume topics starting with `"TOPIC_PREFIX"`:
```
Comment (gaelreyrol) [inline deploy/addon/pulsar.md:156]:
```suggestion
The pulsar add-on given biscuit token can be attenuated, here is an attenuation example using (biscuit-cli)[https://github.com/biscuit-auth/biscuit-cli] from the given biscuit token to produce/consume topics starting with `"TOPIC_PREFIX"`:
```

File: deploy/addon/pulsar.md [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+
+First inspect your biscuit:
+```bash
+$ biscuit inspect addon.biscuit
```
Comment (gaelreyrol) [inline deploy/addon/pulsar.md:160]:
```suggestion
biscuit inspect addon.biscuit
```

File: deploy/addon/pulsar.md [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+==========
+```
+
+* The authority block is the cluster authentication block (the cluster admin biscuit).
+* The block n°1 is an attenuation of the authority block to only authorize operations on `tenant = "user_1235678-f54e-4e09-848c-1953af6e3e89"` and `namespace = "pulsar_1235678-6b36-4af2-be1f-d97862c0c41c"`.
+
+Now attenuate it:
+
+```bash
+$ biscuit attenuate addon.biscuit
```
Comment (gaelreyrol) [inline deploy/addon/pulsar.md:188]:
```suggestion
biscuit attenuate addon.biscuit
```

File: deploy/addon/pulsar.md (hunk 9 above)
Comment (gaelreyrol) [reply deploy/addon/pulsar.md:156]: I think you should give an example to "TOPIC_PREFIX" so we don't assume it is an environment variable.

File: deploy/addon/pulsar.md [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+
+Now the block n°2 ensure that topics must starts with `"TOPIC_PREFIX"`.
+
+You can find more examples on the (biscuit-pulsar authorization java tests)[https://github.com/CleverCloud/biscuit-pulsar/blob/master/src/test/java/com/clevercloud/biscuitpulsar/AuthorizationProviderBiscuitTest.java].
+
+## Storage
+
+### Retention
+
+Provided addon pulsar have a infinite retention policies which can be changed using:
```
Comment (gaelreyrol) [inline deploy/addon/pulsar.md:241]: What are the impacts of an infinite retention policy on Clever Cloud ?
```suggestion
Provided add-on pulsar have an infinite retention policies which can be changed using:
```

File: deploy/addon/pulsar.md [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+```bash
+# Example to set retention of namespace to 2 weeks and/or 100 GB
+pulsarctl --admin-service-url https://<PULSAR_REST_URL> --auth-params <BISCUIT> --auth-plugin org.apache.pulsar.client.impl.auth.AuthenticationToken namespaces set-retention <tenant>/<namespace> --time 2w --size 100G
+```
+
+### Offload storage to Cellar (S3)
+
+Pulsar has a [tiered storage feature](https://pulsar.apache.org/docs/en/tiered-storage-overview/) allowing to offload heavy data to cold storage once a threshold is reached.
+
+For each pulsar addon we provide, we also provide an hidden cellar addon (which is our object storage addon) which is directly binded to the pulsar namespace offload policies.
```
Comment (gaelreyrol) [inline deploy/addon/pulsar.md:252]:
```suggestion
For each pulsar add-on we provide, we also provide a hidden Cellar add-on (which is our object storage add-on) which is directly binded to the pulsar namespace offload policies. 
```
You should provide a link to the Cellar add-on page.

File: deploy/addon/pulsar.md [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+* C++ client
+* C# client
+* Go client
+* Java client
+* Node.js client
+* Python client
+* Rust client
+* WebSocket client
+
+### Operations
```
Comment (gaelreyrol) [inline deploy/addon/pulsar.md:68]: These operations might change in the future ?

File: deploy/addon/pulsar.md [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+```
+PARTITION_READ
+PARTITION_WRITE
+```
+
+### Attenuation
+
+The pulsar addon given biscuit can be attenuated, here is an attenuation example using (biscuit-cli)[https://github.com/biscuit-auth/biscuit-cli] from the given biscuit to produce/consume topics starting with `"TOPIC_PREFIX"`:
+
+First inspect your biscuit:
```
Comment (gaelreyrol) [inline deploy/addon/pulsar.md:158]: We might need to tell how to create this file ^^.

**Review by gaelreyrol — CHANGES_REQUESTED** [review]
I think we should stay consistent about Pulsar with a `P` and an add-on with a `-`.
Biscuit is the engine, so we should write `Biscuit` with a `B` and its token: `Biscuit token`.
We should also use the add-on environment variables in all the examples.

Could you revert the format you've made on the `reference-environmemt-variables.md` file, it is unrelated to this PR and unreviewable ^^.

File: deploy/addon/pulsar.md (hunk 14 above)
Comment (KannarFr) [reply deploy/addon/pulsar.md:70]: yes

File: deploy/addon/pulsar.md (hunk 4 above)
Comment (KannarFr) [reply deploy/addon/pulsar.md:30]: subscription type list, I added clear title.

**Comment by KannarFr** [comment]
I grouped the suggestions and more related changes in f1317cb9c3032383091a429715fd7670838ee875.

About format, it's already one commit: 2d9225f905febfe313249498a272f9fbacefc99e. I also rebased all cleanly so it can be merge without squash and no problem for the format as it's a commit itself. Is that ok for you?

About the version, I am not sure it's a good to precise it in the documentation as we should manage it each time we bump the version.

**Comment by KannarFr** [comment]
About the release pulsar pricing component, @Galimede can you do it if it's ready?

File: deploy/addon/pulsar.md (hunk 14 above)
Comment (gaelreyrol) [reply deploy/addon/pulsar.md:70]: Could you write something about it ? Our customers can reach us through the support to ask new operations ?

File: deploy/addon/pulsar.md (hunk 3 above)
Comment (gaelreyrol) [reply deploy/addon/pulsar.md:12]: As discussed with @cnivolle, we need to have `rabbitmq` & `kafka` keywords around Pulsar ;)

**Review by gaelreyrol — CHANGES_REQUESTED** [review]
It still misses out a few things but it's much better ! Thanks :+1:

**Comment by gaelreyrol** [comment]
> About the version, I am not sure it's a good to precise it in the documentation as we should manage it each time we bump the version.
Could you at least, write some words about it ? What is the release process ? How can they get the version themselves ? I am already anticipating customers reaching the support to ask about the version ^^

File: deploy/addon/pulsar.md [hunk 16]
```diff
@@ -0,0 +1,268 @@
...
+- queue
+- hot storage
+- biscuit
+- rabbitmq
+- kafka
+---
+
+## Overview
+
+Pulsar works on a publisher/subscriber model, just like RabbitMQ or Kafka.
```
Comment (PierreZ) [inline deploy/addon/pulsar.md:19]: Could you add a link to Pulsar's website?
```suggestion
[Pulsar](https://pulsar.apache.org/) works on a publisher/subscriber model, just like RabbitMQ or Kafka.
```

File: deploy/addon/pulsar.md [hunk 17]
```diff
@@ -0,0 +1,268 @@
...
+
```
Comment (PierreZ) [inline deploy/addon/pulsar.md:20]: Before talking consuming strategies, we could talk on:

* why are we using pub/sub tech? 
* why do we love event-driven architecture?

File: deploy/addon/pulsar.md [hunk 18]
```diff
@@ -0,0 +1,268 @@
...
+Topics policies operations authorized:
+
+```
+PARTITION_READ
+PARTITION_WRITE
+```
+
+### Attenuation
+
+The Pulsar add-on given Biscuit token can be attenuated, here is an attenuation example using (biscuit-cli)[https://github.com/biscuit-auth/biscuit-cli] from the given Biscuit token to produce/consume topics starting with `"TOPIC_PREFIX"`:
```
Comment (PierreZ) [inline deploy/addon/pulsar.md:158]: Could you add an attenuation example?

For example, "I want to derive a biscuit from the one I have that can only read data"

**Review by PierreZ — CHANGES_REQUESTED** [review]
Nice work @Keksoj! I added a few comments.

File: deploy/addon/pulsar.md (hunk 18 above)
Comment (Keksoj) [reply deploy/addon/pulsar.md]: I think the attenuation example is already provided.

**Comment by Keksoj** [comment]
Okay, I :

- reverted the formatting of the environment variable reference
- changed all generic variable mentions, like <BISCUIT>, to the actual env. variables, like $PULSAR_ADDON_TOKEN
- changed the prefix in the example to a more understandable syntax

Yet to do: 

- [ ] add some context to why we like pub/sub and event driven architecture
- [ ] pricing 
- plus some other things listed accross the PR

**Comment by gaelreyrol** [comment]
> Yet to do:
>     * [ ]  pricing

I just created a pull request to update cellar and fsbucket plans: https://github.com/CleverCloud/doc.clever-cloud.com/pull/662.
You can use the same shortcode with `pulsar` like so: `{{< pricingConsumption "pulsar" >}}`.

File: deploy/addon/pulsar.md (hunk 17 above)
Comment (KannarFr) [reply deploy/addon/pulsar.md:20]: @PierreZ Can you contribute these points?

File: deploy/addon/pulsar.md (hunk 18 above)
Comment (KannarFr) [reply deploy/addon/pulsar.md]: It is.

**Comment by KannarFr** [comment]
@Keksoj you are using an old pulsar, the correct env var is `ADDON_PULSAR_TOKEN`, can you fix it? Delete your current addon and create a new one. Sorry about that.

File: deploy/addon/pulsar.md (hunk 17 above)
Comment (KannarFr) [reply deploy/addon/pulsar.md:20]: I've done it on 542d67ed3e05f37baf5ca9b17339323f153998b4 and 1ae3ab7a3b8252957823202075500eed6067c30c.

**Comment by Keksoj** [comment]
@KannarFr sure !

**Comment by Keksoj** [comment]
ping me when you're done, I've got some cosmetic changes to do.

**Comment by KannarFr** [comment]
@Keksoj I did it.

@gaelreyrol can you rereview and eventually merge? Then I will bump the related blogpost.

**Comment by Keksoj** [comment]
@KannarFr yup!
I will ping @gaelreyrol when done

**Comment by KannarFr** [comment]
Oh, the Authorization with Attenuation part must be before the Usage IMHO.

**Comment by Keksoj** [comment]
@gaelreyrol Ready for review!

**Review by gaelreyrol — CHANGES_REQUESTED** [review]
One last thing, you should only use single `\` in your bash examples.

**Comment by BlackYoup** [comment]
I'll review this at some point, but don't wait for my review if this is blocking. Just a question: should we start adding CLI references here and there in the documentation? For example, in the `Create add-on` section, we could add `clever addon create pulsar ....`.

I'm starting to write the Jenkins documentation and though about this.

**Comment by BlackYoup** [comment]
> I'll review this at some point, but don't wait for my review if this is blocking. Just a question: should we start adding CLI references here and there in the documentation? For example, in the `Create add-on` section, we could add `clever addon create pulsar ....`.
> 
> I'm starting to write the Jenkins documentation and though about this.

I just saw that there is a partial `create-application.md` that display a CLI command. Maybe we could have the same type of partial for add-ons. What do you think?

**Comment by gaelreyrol** [comment]
> > I'll review this at some point, but don't wait for my review if this is blocking. Just a question: should we start adding CLI references here and there in the documentation? For example, in the `Create add-on` section, we could add `clever addon create pulsar ....`.
> > I'm starting to write the Jenkins documentation and though about this.
> 
> I just saw that there is a partial `create-application.md` that display a CLI command. Maybe we could have the same type of partial for add-ons. What do you think?

That's a great idea ! :slightly_smiling_face:

**Comment by PierreZ** [comment]
Nice work :+1: :rocket:

File: deploy/addon/pulsar.md [hunk 19]
```diff
@@ -0,0 +1,413 @@
...
+## Common use cases
+
+* Ingestion user interaction and server events To make use of user interaction events from end-user apps or server events from your system, you may forward them to Pulsar and then use a stream processing tool which delivers them to your applications. Pulsar allows you to gather events from many clients simultaneously.
+* Replicating data among databases using [Pulsar IO](https://pulsar.apache.org/docs/en/io-overview/) is commonly used to distribute change events from databases.
+* Parallel processing and workflows. You can efficiently distribute a large number of tasks among multiple workers ( compressing text files, sending email notifications).
+* Data streaming from IoT devices. For example, a residential sensor can stream data to backend servers.
+* Refreshing distributed caches. For example, an application can publish invalidation events to update the IDs of objects that have changed.
+* Real-time event distribution Events, raw or processed, may be made available to multiple applications across your team and organization for real time processing.
+
+## Create an add-on
```
Comment (BlackYoup) [inline deploy/addon/pulsar.md:61]:
```suggestion
## Create a Pulsar add-on
```

File: deploy/addon/pulsar.md [hunk 20]
```diff
@@ -0,0 +1,413 @@
...
+
+`{persistent|non-persistent}://<CLEVERCLOUD_TENANT_ID>/<ADDON_ID>/<TOPIC_NAME>`
+
+## Version
+
+We maintain up-to-date Pulsar clusters based on the official Apache Pulsar release process. Your Pulsar add-on version is available in your add-on dashboard. Updates will be notified using Clever Cloud weekly updates on the [blog](https://www.clever-cloud.com/blog/).
+
+## Common use cases
+
+* Ingestion user interaction and server events To make use of user interaction events from end-user apps or server events from your system, you may forward them to Pulsar and then use a stream processing tool which delivers them to your applications. Pulsar allows you to gather events from many clients simultaneously.
```
Comment (BlackYoup) [inline deploy/addon/pulsar.md:54]: I can't really understand this sentence, am I the only one?

File: deploy/addon/pulsar.md [hunk 21]
```diff
@@ -0,0 +1,413 @@
...
+
+More on this in the [official documentation](https://pulsar.apache.org/docs/en/concepts-messaging/#subscriptions).
+
+A topic is defined this way:
+
+`{persistent|non-persistent}://tenant/namespace/topic`
+
+Tenants and namespaces allow for grouping and subgrouping of topics.
+
+A Clever Cloud Pulsar add-on is basically a immutable `tenant/namespace` where the tenant is your user id, and the namespace is the add-on id.
```
Comment (BlackYoup) [inline deploy/addon/pulsar.md:43]: `an immutable`?

File: deploy/addon/pulsar.md [hunk 22]
```diff
@@ -0,0 +1,413 @@
...
+
+## Version
+
+We maintain up-to-date Pulsar clusters based on the official Apache Pulsar release process. Your Pulsar add-on version is available in your add-on dashboard. Updates will be notified using Clever Cloud weekly updates on the [blog](https://www.clever-cloud.com/blog/).
+
+## Common use cases
+
+* Ingestion user interaction and server events To make use of user interaction events from end-user apps or server events from your system, you may forward them to Pulsar and then use a stream processing tool which delivers them to your applications. Pulsar allows you to gather events from many clients simultaneously.
+* Replicating data among databases using [Pulsar IO](https://pulsar.apache.org/docs/en/io-overview/) is commonly used to distribute change events from databases.
+* Parallel processing and workflows. You can efficiently distribute a large number of tasks among multiple workers ( compressing text files, sending email notifications).
```
Comment (BlackYoup) [inline deploy/addon/pulsar.md:56]:
```suggestion
* Parallel processing and workflows. You can efficiently distribute a large number of tasks among multiple workers (compressing text files, sending email notifications).
```

File: deploy/addon/pulsar.md [hunk 23]
```diff
@@ -0,0 +1,413 @@
...
+A Clever Cloud Pulsar add-on is basically a immutable `tenant/namespace` where the tenant is your user id, and the namespace is the add-on id.
+It allows you to create and use topics following this pattern:
+
+`{persistent|non-persistent}://<CLEVERCLOUD_TENANT_ID>/<ADDON_ID>/<TOPIC_NAME>`
+
+## Version
+
+We maintain up-to-date Pulsar clusters based on the official Apache Pulsar release process. Your Pulsar add-on version is available in your add-on dashboard. Updates will be notified using Clever Cloud weekly updates on the [blog](https://www.clever-cloud.com/blog/).
+
+## Common use cases
```
Comment (BlackYoup) [inline deploy/addon/pulsar.md:52]: Almost of the items of the list use a different "example" pattern. Maybe make all the elements like:
`- <example>: explain it`

**Comment by KannarFr** [comment]
> One last thing, you should only use single `\` in your bash examples.

fixed in 17280cd860b2085f150946451291262a2d9310ad.

@BlackYoup I agree with you, as we need to release the blogpost fast, can you file an issue?

@gaelreyrol can you merge, release and publish this?

File: deploy/addon/pulsar.md [hunk 24]
```diff
@@ -0,0 +1,413 @@
...
+In your personnal space, click on *Create* > *an add-on* > *Pulsar*.
+Choose your plan, link an app to it (or not), give it a name and a zone, and it's done.
+
+## Authorization
+
+Pulsar add-on uses [Biscuit for Pulsar](https://github.com/CleverCloud/biscuit-pulsar) implementation which is directly pluggable to the Pulsar authentication and authorization system. Each add-on exposes its own Biscuit token.
+
+### Attenuation
+
+The Pulsar add-on given Biscuit token can be attenuated, here is an attenuation example using [biscuit-cli](https://github.com/biscuit-auth/biscuit-cli) from the given Biscuit token to produce/consume topics starting with a custom topic prefix called `"my-own-prefix"`.
```
Comment (BlackYoup) [inline deploy/addon/pulsar.md:73]:
```suggestion
The Biscuit provided by a pulsar add-on can be attenuated. Here is an attenuation example using [biscuit-cli](https://github.com/biscuit-auth/biscuit-cli) to produce/consume topics starting with a custom topic prefix called `"my-own-prefix"`.
```

Also, should Biscuit put between back ticks? `Biscuit`

File: deploy/addon/pulsar.md [hunk 25]
```diff
@@ -0,0 +1,413 @@
...
+
+Put your Biscuit token in a file:
```
Comment (BlackYoup) [inline deploy/addon/pulsar.md:75]: Write your biscuit token?

File: deploy/addon/pulsar.md [hunk 26]
```diff
@@ -0,0 +1,413 @@
...
+check if topic_operation(#ambient, "user_12345678-f54e-4e09-848c-1953af6e3e89", "pulsar_12345678-6b36-4af2-be1f-d97862c0c41c", $topic, $operation), $topic.starts_with("my-own-prefix");
+
+== Revocation ids ==
+Content-based: 8eaaa639d5b94c3e053ad840e8dcca6fe66a621442ecc99eadf1df03d6138f1d
+Unique:        f608dc2f724fc14faf0daf50774ef0b9425cda26f56ee93317ca80ca13736027
+
+==========
+```
+
+Now the block n°2 ensure that topics must starts with `"my-own-prefix"`.
```
Comment (BlackYoup) [inline deploy/addon/pulsar.md:157]:
```suggestion
Now the block n°2 ensures that topics must start with `"my-own-prefix"`.
```

File: deploy/addon/pulsar.md [hunk 27]
```diff
@@ -0,0 +1,413 @@
...
+We advise you to use [`pulsarctl`](https://github.com/streamnative/pulsarctl) provided by StreamNative. Here is an example to list topics in your add-on (in your namespace):
+
+```bash
+pulsarctl --admin-service-url $ADDON_PULSAR_HTTP_URL \\
+          --auth-params $ADDON_PULSAR_TOKEN \\
+          --auth-plugin org.apache.pulsar.client.impl.auth.AuthenticationToken \\
+          topics list $ADDON_PULSAR_TENANT/$ADDON_PULSAR_NAMESPACE
+```
+
+As Biscuit is a token, you can use `AuthenticationToken($ADDON_PULSAR_TOKEN)` provided by [clients libraries](https://pulsar.apache.org/docs/en/client-libraries/) to authenticate to our clusters without any tweak.
```
Comment (BlackYoup) [inline deploy/addon/pulsar.md:172]: I think this should go at the end of the `## Authorization` section, before the `### Attenuation` one

File: deploy/addon/pulsar.md [hunk 28]
```diff
@@ -0,0 +1,413 @@
...
+
+while (!consumer.hasReachedEndOfTopic()) {
+  Message<String> msg = consumer.receive();
+  // Got the message!
+}
+```
+
+### Operations
+
+The pulsar addon given Biscuit token enables you to run several operations on namespace, its policies and the related topics.
```
Comment (BlackYoup) [inline deploy/addon/pulsar.md:293]:
```suggestion
The Biscuit token provided by the Pulsar add-on allows you to run several operations on namespace, its policies and the related topics.
```

File: deploy/addon/pulsar.md [hunk 29]
```diff
@@ -0,0 +1,413 @@
...
+
+These operations might change in the future. Don't hesitate to call our support to ask for new operations!
```
Comment (BlackYoup) [inline deploy/addon/pulsar.md:295]:
```suggestion
These operations might change in the future. Don't hesitate to write to our support to ask for new operations!
```

File: deploy/addon/pulsar.md [hunk 30]
```diff
@@ -0,0 +1,413 @@
...
+# Example to set retention of namespace to 2 weeks and/or 100 GB
+pulsarctl --admin-service-url $ADDON_PULSAR_HTTP_URL \\
+          --auth-params $ADDON_PULSAR_TOKEN \\
+          --auth-plugin org.apache.pulsar.client.impl.auth.AuthenticationToken \\
+          namespaces set-retention $ADDON_PULSAR_TENANT/$ADDON_PULSAR_NAMESPACE --time 2w --size 100G
+```
+
+### Offload storage to Cellar (S3)
+
+Pulsar has a [tiered storage feature](https://pulsar.apache.org/docs/en/tiered-storage-overview/) allowing to offload heavy data to cold storage once a threshold is reached.
```
Comment (BlackYoup) [inline deploy/addon/pulsar.md:395]:
```suggestion
Pulsar has a [tiered storage feature](https://pulsar.apache.org/docs/en/tiered-storage-overview/) allowing offload of heavy data to cold storage once a threshold is reached.
```

File: deploy/addon/pulsar.md [hunk 31]
```diff
@@ -0,0 +1,413 @@
...
+
+For each Pulsar add-on we provide, we also provide an hidden [Cellar add-on](https://www.clever-cloud.com/doc/deploy/addon/cellar/) (which is our object storage addon) which is directly binded to the Pulsar namespace offload policies.
```
Comment (BlackYoup) [inline deploy/addon/pulsar.md:397]:
```suggestion
For each Pulsar add-on we provide, we also provide a hidden [Cellar add-on]({{< ref "deploy/addon/cellar.md" >}}), our object storage add-on. This Cellar add-on is bound to the Pulsar namespace offload policies and will store your offloaded data.
```

**Review by BlackYoup — CHANGES_REQUESTED** [review]
Let me know what you think of those requests :)

File: deploy/addon/pulsar.md (hunk 23 above)
Comment (KannarFr) [reply deploy/addon/pulsar.md:52]: Well no, it is too complicated to make generic as some are parts of other etc.

**Comment by KannarFr** [comment]
@BlackYoup I applied almost all of your suggestions. @gaelreyrol gogo! :p

**Review by gaelreyrol — APPROVED** [review]

**Comment by gaelreyrol** [comment]
Nice work guys ! :clap:

**Pull request merged**